sftp/sftp.go

```diff
@@ -101,6 +96,77 @@ func (fm *fileManagerImpl) Delete(remoteFilePath string) error {
 	if err != nil {
 		return fmt.Errorf("cannot delete file: %w", err)
 	}
+	return nil
+}
 
+func (fm *fileManagerImpl) Reset() error {
+	newClient, err := newSFTPClientFromConfig(fm.config)
+	if err != nil {
+		return err
+	}
+	fm.client = newClient
 	return nil
 }
+
+func NewFileManager(config *SSHConfig) (FileManager, error) {
+	sftpClient, err := newSFTPClientFromConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &fileManagerImpl{client: sftpClient, config: config}, nil
+}
+
+type retryableFileManagerImpl struct {
+	fileManager FileManager
+}
+
+func (r *retryableFileManagerImpl) Upload(localFilePath, remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Upload(localFilePath, remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Download(remoteFilePath, localDir string) error {
+	fileOperation := func() error {
+		return r.fileManager.Download(remoteFilePath, localDir)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Delete(remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Delete(remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
```

Comment:
Are you sure this is the right approach? What if the connection is lost in the middle of the `Delete` operation? 

Is the `Delete` operation idempotent?

```go
// Remove removes the specified file or directory. An error will be returned if no
// file or directory with the specified path exists, or if the specified directory
// is not empty.
func (c *Client) Remove(path string) error {
```

What happens if we issue a delete, the file is deleted but before the client can return a response the connection is lost. We would retry and fail with an error even though the file was correctly deleted.

Reply:
By the way, wouldn't it make sense to have the retries on the `SSHClient` instead of having them in the `FileManager`? This way you don't retry the whole thing (since upload and download do multiple things) but you can retry only the last thing that didn't succeed. Wdyt?

Reply:
The reason for retrying on the SFTP client (part of the FileManager) instead of the SSHClient is: 
- A connection lost error requires creating a new connection. Simply creating a new SSHClient would not suffice; we would also need to create a new SFTP client using the existing SSHClient with sftp.NewClient(sshClient).

Reply:
https://github.com/rudderlabs/rudder-go-kit/pull/465#discussion_r1604861526
- This connection lost is different. The retry we implemented is more specific to `"retry on idle connections"`. We can rename the functions and variables for better clarity.

Reply:
@fracasula 
https://github.com/rudderlabs/rudder-go-kit/pull/465#discussion_r1604861526

This connection lost is different. The retry we implemented is more specific to "retry on idle connections". We can rename the functions and variables for better clarity.
cc- @koladilip